proposals/intraprocess_communication.md

```diff
@@ -0,0 +1,580 @@
+the program when standard intra-process communication mechanism is used. Since the experiments have been run
+for 120 seconds, there is an increase of approximately 60KB per second. However, even considering the initial
+memory usage, it is possible to see how it is affected from the presence of the additional publishers and
+subscriptions needed for intra-process communication. There is a difference of 10MB in Sierra Nevada and of
+33MB in Mont Blanc between standard intra-process communication on and off.
+
+The last experiment show how the current implementation performs in the case that both intra and
+inter-process communication are needed. The test consists of running Sierra Nevada on RaspberryPi 2, and,
+in a separate desktop machine, a single node subscribing to all the available topics coming from Sierra Nevada.
+This use-case is common when using tools such as `rosbag` or `rviz`.
+
+| ROS2 system                   |  IPC     |    RMW        | Latency [us] | CPU [%] | RAM [Mb] |
+| -------------                 |  -----   | ------------- | ------------ | ------- | -------- |
+| Sierra Nevada + debug node    |   off    | Fast-RTPS     |      800     |   22    |    50    |
+| Sierra Nevada + debug node    | standard | Fast-RTPS     |     1100     |   35    |  60->65  |
+| Sierra Nevada + debug node    |   new    | Fast-RTPS     |      180     |   15    |    32    |
+
+
+These results show that if there is at least one node in a different process, with the current implementation
+it is better to keep intra-process communication disabled. The proposed implementation does not require the
+ROS2 middleware when publishing intra-process. This allows to easily remove the connections between nodes in
+the same process when it is required to publish also inter process, potentially resulting in a very small
+overhead with respect to the only intra-process case.
+
+
+## Open Issues and Future Works
+
+There are some open issues that are not addressed neither on the current implementation nor on the
+proposed one.
+
+ - The proposal does not take into account the problem of having a queue with twice the size when both inter
+   and intra-process communication are used. A `Publisher` or a `Subscription` with a history depth of 10 will
+   be able to store up to 20 messages without processing them (10 intra-process and 10 inter-process). This
+   issue is also present in the current implementation, since each `Subscription` is doubled.
+
+ - The proposal does not allow to handle the scenario in which a `transient local` `Publisher` has only
+   intra-process `Subscription`s when it is created, but, eventually, a `transient local` `Subscription` in a
+   different process joins. Initially, published messages are not passed to the middleware, since all the
+   `Subscription`s are in the same process. This means that the middleware is not able to store old messages
+   for eventual late-joiners.
```

Comment:
That is a good example for why a transient local publisher *must* publish to the RMW impl even if at the time of publishing no inter process subscribers exist.

Reply:
Agreed.